src/thunderbots/software/geom/util.cpp

```diff
@@ -0,0 +1,914 @@
+    }
+
+    return false;
+}
+
+bool contains(const Rect &out, const Vector &in)
+{
+    return out.containsPoint(in);
+}
+
+bool intersects(const Triangle &first, const Circle &second)
+{
+    return contains(first, second.origin) ||
+           dist(getSide(first, 0), second.origin) < second.radius ||
+           dist(getSide(first, 1), second.origin) < second.radius ||
+           dist(getSide(first, 2), second.origin) < second.radius;
+}
+bool intersects(const Circle &first, const Triangle &second)
+{
+    return intersects(second, first);
+}
+
+bool intersects(const Circle &first, const Circle &second)
+{
+    return (first.origin - second.origin).len() < (first.radius + second.radius);
+}
+
+bool intersects(const Ray &first, const Seg &second)
+{
+    if (std::abs(first.toVector().cross(second.toVector())) > EPS)
+    {
+        Vector isect = lineIntersection(first.start, first.dir, second.start, second.end);
+        return contains(first, isect) && contains(second, isect);
+    }
+    return collinear(first.start, first.dir, second.start);
+}
+bool intersects(const Seg &first, const Ray &second)
+{
+    return intersects(second, first);
+}
+
+bool intersects(const Seg &first, const Circle &second)
+{
+    // if the segment is inside the circle AND at least one of the points is
+    // outside the circle
+    return contains(second, first) &&
+           (distsq(first.start, second.origin) > second.radius * second.radius ||
+            distsq(first.end, second.origin) > second.radius * second.radius);
+}
+bool intersects(const Circle &first, const Seg &second)
+{
+    return intersects(second, first);
+}
+
+bool intersects(const Seg &first, const Seg &second)
+{
+    if (sign((first.start - first.end).cross(second.start - second.end)) == 0)
+    {
+        // find distance of two endpoints on segments furthest away from each
+        // other
+        double mx_len = std::sqrt(std::max(
+            std::max(
+                (second.start - first.end).lensq(), (second.end - first.end).lensq()),
+            std::max(
+                (second.start - first.start).lensq(),
+                (second.end - first.start).lensq())));
+        // if the segments cross then this distance should be less than
+        // the sum of the distances of the line segments
+        return mx_len <
+               (first.start - first.end).len() + (second.start - second.end).len() + EPS;
+    }
+
+    return sign((first.end - first.start).cross(second.start - first.start)) *
+                   sign((first.end - first.start).cross(second.end - first.start)) <=
+               0 &&
+           sign((second.end - second.start).cross(first.start - second.start)) *
+                   sign((second.end - second.start).cross(first.end - second.start)) <=
+               0;
+}
+
+template <size_t N>
+Vector getVertex(const Poly<N> &poly, unsigned int i)
+{
+    return poly[i % N];
```

Comment:
Instead of modulo, should probably clamp index to [0, N].

Reply:
bump

Reply:
bump